packages/ckeditor5-widget/src/utils.js

```diff
@@ -339,6 +342,75 @@ export function viewToModelPositionOutsideModelElement( model, viewElementMatche
+ *	│  ┃   ╰─────────────────╯   ┃  │   │  ┃                         ┃  │   │  ┃                         ┃  │
+ *	│  ┃                         ┃  │   │  ┃                         ┃  │   │  ┃                         ┃  │
+ *	│  ┃                         ┃  │   │  ┃                         ┃  │   │  ┃   ╭─────────────────╮   ┃  │
+ *	│  ┃                         ┃  │   │  ┃                         ┃  │   │  ┃   │     Balloon     │   ┃  │
+ *	│  ┃                         ┃  │   │  ┃                         ┃  │   │  ┃   ╰───────\ /───────╯   ┃  │
+ *	│  ┃                         ┃  │   │  ┃                         ┃  │   │  ┃            V            ┃  │
+ *	│  ┃                         ┃  │   │  ┃                         ┃  │   │  ┗━━━━━━━━━━━━━━━━━━━━━━━━━┛  │
+ *	└──╀─────────────────────────╀──┘   └──╀─────────────────────────╀──┘   └───────────────────────────────┘
+ *	   ┊                         ┊         ┊                         ┊
+ *	   ┊                         ┊         └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
+ *	   ┊                         ┊
+ *	   └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
+ *
+ * **Note**: Works best if used together with
+ * {@link module:ui/panel/balloon/balloonpanelview~BalloonPanelView.defaultPositions default `BalloonPanelView` positions}
+ * like `northArrowSouth` and `southArrowNorth`; the transition between these two and this position is smooth.
+ *
+ * @param {module:utils/dom/rect~Rect} widgetRect A rect of the widget.
+ * @param {module:utils/dom/rect~Rect} balloonRect A rect of the balloon.
+ * @returns {module:utils/dom/position~Position}
+ */
+export function centeredBalloonPositionForLongWidgets( widgetRect, balloonRect ) {
+	const viewportRect = new Rect( global.window );
+	const viewportWidgetInsersectionRect = viewportRect.getIntersection( widgetRect );
+	const isUpperTargetEdgeOffViewport = widgetRect.top < viewportRect.top;
+	const isLowerTargetEdgeOffViewport = widgetRect.bottom > viewportRect.bottom;
+
+	// Because this is a last resort positioning, to keep things simple we're not playing with positions of the arrow
+	// like, for instance, "south west" or whatever. Just try to keep the balloon in the middle of the visible area of
+	// the widget for as long as it is possible. If the widgets becomes invisible (because cropped by the viewport),
+	// just... place the balloon in the middle of it (because why not?).
+	const targetRect = viewportWidgetInsersectionRect || widgetRect;
+	const left = targetRect.left + targetRect.width / 2 - balloonRect.width / 2;
+
+	// Case (C).
+	if ( isUpperTargetEdgeOffViewport && !isLowerTargetEdgeOffViewport ) {
+		return {
+			top: Math.min( widgetRect.bottom, viewportRect.bottom ) - balloonRect.height - BalloonPanelView.arrowVerticalOffset,
+			left,
+			name: 'arrow_s'
+		};
+	}
+	// Cases (A) and (B).
+	else {
+		return {
+			top: Math.max( widgetRect.top, 0 ) + BalloonPanelView.arrowVerticalOffset,
+			left,
+			name: 'arrow_n'
+		};
+	}
```

Comment:
I would simplify this function only to cases A and B. C seems to appear only in very "short" scenario and makes balloon jump on the bottom of the widget.
```suggestion
	return {
		top: Math.max( widgetRect.top, 0 ) + BalloonPanelView.arrowVerticalOffset,
		left,
		name: 'arrow_n'
	};
```

Reply:
I also had some doubts about it but I figured it feels sort of OK when the balloon is attached to the bottom and there's not enough space to keep it there. This way we avoid a rapid change and provide this one extra step `[ underneath, inside at the bottom, inside at the top]` instead of `[underneath, inside at the top]`. 

Anyway, a funny idea popped into my head when I was thinking about it. What if... the position detects which direction the window is scrolling? Like as long as scrolling down (the upper edge of widget disappears) we would attach it to the top of the viewport and to the bottom of the viewport if scrolling in to opposite direction. 

WDYT?

Reply:
As we talked F2F, for this edge scenario there is no point in complicating it.